![logo](transgram.png)
Description
================

TransGram is a genome-guided transcriptome reconstruction tool designed for long RNA-seq reads. It first analyzes the data heterogeneity, classifies the input into two categories (Alpha and Beta), and then applies a tailored splice-graph-based path reconstruction strategy accordingly. In addition, a machine learning model is incorporated for quality assessment and filtering of the reconstructed transcripts.

TransGram supports two reconstruction modes:
	•	Annotation-free mode — requires only a sorted BAM file as input.
	•	Annotation-guided mode — utilizes a user-provided annotation file to assist reconstruction, requiring both a sorted BAM file and an annotation file.

Prerequisites
================

  g++ with support for C++11 (e.g. 9.4.0)
  
  Please ensure that [xgboost][xgboost] is properly installed !
 

# 1. Installing and test

    
    (A) Install XGBoost (v2.1.3), scikit-learn (v1.7.1), numpy (v1.26.4), and pandas (v2.2.2).
   
       Using pip:
          $ pip install xgboost==2.1.3 scikit-learn==1.7.1 numpy==1.26.4 pandas==2.2.2
       Using conda:
          $ conda install -c conda-forge xgboost=2.1.3 scikit-learn=1.7.1 numpy=1.26.4 pandas=2.2.2
          
    (B) Download and build TransGram.
    
          $ git clone https://github.com/yutingsdu/TransGram 
          $ cd TransGram
          $ make all release
          
	      Completing the installation procedure will require a few minutes. 
	  
    (C) Test TransGram on the demo dataset.
        
        Change to the TransGram/sample_test/ directory, and run:
        
          $ ./run_me.sh
          
	      If the file transgram_outdir/TransGram.gtf is generated, congratulations—you have successfully installed TransGram.
      
        

# 2. Usage 


TransGram-v.0.0 usage:

* Required *

For training a customized model, please refer to the [customized model][customized model] section.

    --bam/-b <string>		: Path to the alignment BAM files generated by minimap2.

---------------------------------------------------------------------------

* Options *

    --help/-h			: Output TransGram Help Information

    --version/-v		: Print current version of TransGram

    --annotation/-G	       <string>	: Reference annotation to use for guiding the assembly process. (GTF)

    --specify_data_type/-S <string>	: Specify data type to be "alpha" or "beta". (default: data type will be determined by TransGram)

    --output_dir/-o        <string>	: Output path for TransGram (default:transgram_outdir)

    --min_trans_length/-l  <int>   	: Minimum assembled transcript length (default:200)

    --min_trans_score/-s   <double> 	: Minimum xGboost score for the assembled transcripts (default:0.1)

    --min_average_frac/-c  <double>  : Minimum junction coverage fraction by average junction coverage (default:0.001)

    --QA		           <string>	: Path to the transcripts generated from the correspoding BAM file (in GTF format).
				        TransGram will assess the quality and filter them based on a specific parameter (requires -s)

    --train 			: Train a new modle using user-provided data.
				        (Requires the -b, -G, and -o  option, and gffcompare must be installed.)

    --CusModel	           <string>  : Path to the customized model created by the users.

---------------------------------------------------------------------------

* Typical Commands *

A typical TransGram command for the long RNA-seq data under annotation-free mode might be:

    TransGram -b alignment.bam -o transgram_outdir
    
A command for the long RNA-seq data under annotation-guided mode might be:
    
    TransGram -b alignment.bam -G annotation.gtf -o transgram_outdir

Perform quality assessment on your transcripts (my-transcript.gtf) using TransGram.

    TransGram -b alignment.bam --QA my-transcript.gtf -o transgram_outdir

---------------------------------------------------------------------------

# 3. Output Description 

TransGram generates the following files:
After transcriptome reconstruction, TransGram generates the following files:
	•	TransGram.gtf
Final set of multi-exon transcripts reconstructed by TransGram, filtered by quality score.
	•	TransGram-singleExon.gtf
Final set of single-exon transcripts reconstructed by TransGram.
	•	TransGram-Abundance.tsv
Transcript-level expression estimates for all transcripts in TransGram.gtf and TransGram-singleExon.gtf.

➡️ The above three files are recommended for downstream analysis.
	•	TransGram-temp.gtf
Initial set of all multi-exon transcripts reconstructed by TransGram (before filtering by quality score).
	•	TransGram.score
Machine learning–based quality scores corresponding to the transcripts in TransGram-temp.gtf.

    
The following four files saved in the /tmp directory are temporary intermediates created during splice graph construction from alignment data:
transgram.graph, gtf2path.info, data.info, transgram.reads.raw.align, transgram.read.length


# 4. Running TransGram with a customized model


If you choose to run TransGram with a customized model for filtering the discovered transcripts, please refer to [training][training].

---------------------------------------------------------------------------

Authors: Ting Yu and Zitong Ren designed and wrote TransGram.<br>
Contact:<br>
Any questions, problems, bugs are welcome and should be dumped to Ting Yu <yutingsdu@163.com> <br>
Updated on Dec 13, 2024. <br>

[xgboost]: https://github.com/dmlc/xgboost
[training]: https://github.com/yutingsdu/TransGram/tree/main/xgboost_predict/train
[customized model]:https://github.com/yutingsdu/TransGram?tab=readme-ov-file#3-running-transgram-with-a--customized-model
