Description
================

TransGram is a genome-guided transcriptome reconstruction tool for long RNA-seq reads.

This is a function of TransGram: You can train a new quality assessment model with your sample.
In situations where the default model you though is not suitable for the sample, TransGram provides a function to train a new model. 
To ensure that the trained model captures data features effectively, the data used for training should be well annotated.The new model 
can be used on the same or similar species.

The users should provide a BAM file generated by minimap2, an annotation file to train a new model.


Prerequisites
================

Please ensure that [xgboost][xgboost], [TransGram][TransGram], and [GuffCompare][GuffCompare] are properly installed 
and add the software binary directory to the PATH environment variable in the shell configuration file
!

An illustration of installing the used tools.
================

Installing XGBoost (v2.1.3), scikit-learn (v1.7.1), numpy (v1.26.4), and pandas (v2.2.2).

Installing GffCompare

    $ wget https://ccb.jhu.edu/software/stringtie/dl/gffcompare-0.12.10.Linux_x86_64.tar.gz
    $ tar zxvf gffcompare-0.12.10.tar.gz

Installing TransGram.

    $ git clone https://github.com/yutingsdu/TransGram
    $ cd TransGram
    $ make release


Please don't forget to set the enviroment variable for GffCompare.

OR

Type the command 

export PATH=your-path/GffCompare:$PATH

before running, and /your-path/ is the directory where GffCompare installed.


An example for training a new quality assessment model.
================
    
    TransGram -b sample1.bam -G annotations.gtf -o CusModel_path --train
    
    If everything goes well, you will see a directory "transgram_new_model"

Runing TransGram with the new quality assessment model.
================
    
    TransGram -b alignment.bam -o transgram-outdir --CusModel CusModel_path

[xgboost]: https://github.com/dmlc/xgboost
[TransGram]: https://github.com/yutingsdu/TransGram
[StringTie2]: https://github.com/skovaka/stringtie2 
[GuffCompare]:https://github.com/gpertea/gffcompare?tab=readme-ov-file
[cufflinks]: http://cole-trapnell-lab.github.io/cufflinks/
